transport: service_level_controller: create and use `driver` service level

andrzej-jackowski-scylladb opened the pull request:
This patch series:
 - Increases the number of allowed scheduling groups to allow creation of `sl:driver`
 - Implements `create_driver_service_level` that creates `sl:driver` with shares=200 if it wasn't already created
 - Implements creation of `sl:driver` for new systems and tests in `raft_initialize_discovery_leader`
 - Modifies `topology_coordinator` to use  create `sl:driver` after upgrades.
 - Implements using `sl:driver` for new connections in `transport/server`
 - Adds to `transport/server` recognition of driver's control connections and forcing them to keep using `sl:driver`.
 - Adds tests to verify the new functionality
 - Modifies existing tests to let them pass after `sl:driver` is added
 - Modifies the documentation to contain new `sl:driver`


The changes were evaluated by a test with the following scenario ([test_connections-sl-driver.py](https://github.com/user-attachments/files/22021273/test_connections-sl-driver.py)):
 - Start ScyllaDB with one node
 - Create 1000 keyspaces, 1 table in each keyspace
 - Start `cassandra-stress` (`-rate threads=50  -mode native cql3`)
 - Run connection storm with 1000 session (100 python processes, 10 sessions each)

The maximum latency during connection storm dropped **from 224.94ms to 41.43ms** (those numbers are average from 20 test executions, were max latency was in [140ms, 361ms] before change and [31.4ms, 61.5ms] after).

The snippet of cassandra-stress output from the moment of connection storm:
Before:
```
type       total ops,    op/s,    pk/s,   row/s,    mean,     med,     .95,     .99,    .999,     max,   time,   stderr, errors,  gc: #,  max ms,  sum ms,  sdv ms,      mb
...
total,        789206,   85887,   85887,   85887,     0.6,     0.3,     2.0,     2.0,     2.5,     5.0,    9.0,  0.09679,      0,      0,       0,       0,       0,       0
total,        909322,  120116,  120116,  120116,     0.4,     0.2,     1.9,     2.0,     2.1,     3.1,   10.0,  0.09053,      0,      0,       0,       0,       0,       0
total,        964392,   55070,   55070,   55070,     0.9,     0.4,     2.0,     4.5,     7.7,    18.9,   11.0,  0.09203,      0,      0,       0,       0,       0,       0
total,        975705,   11313,   11313,   11313,     4.4,     3.5,     6.5,    24.5,    82.7,    83.0,   12.0,  0.11713,      0,      0,       0,       0,       0,       0
total,        987548,   11843,   11843,   11843,     4.2,     3.5,     6.5,    33.7,    48.6,    51.5,   13.0,  0.13366,      0,      0,       0,       0,       0,       0
total,        995422,    7874,    7874,    7874,     6.3,     4.0,     7.7,    85.6,   112.9,   113.5,   14.0,  0.14753,      0,      0,       0,       0,       0,       0
total,       1007228,   11806,   11806,   11806,     4.3,     3.5,     6.5,    29.1,    43.8,    87.1,   15.0,  0.15598,      0,      0,       0,       0,       0,       0
total,       1012840,    5612,    5612,    5612,     8.2,     5.0,    11.5,   121.8,   166.6,   170.1,   16.0,  0.16535,      0,      0,       0,       0,       0,       0
total,       1016186,    3346,    3346,    3346,    13.4,     7.4,    20.1,   204.9,   207.6,   210.4,   17.0,  0.17405,      0,      0,       0,       0,       0,       0
total,       1025462,    9276,    9276,    9276,     6.3,     3.9,     9.6,    74.6,   206.8,   210.0,   18.0,  0.17800,      0,      0,       0,       0,       0,       0
total,       1035979,   10517,   10517,   10517,     4.8,     3.5,     6.7,    38.5,    82.6,    83.0,   19.0,  0.18120,      0,      0,       0,       0,       0,       0
total,       1047488,   11509,   11509,   11509,     4.3,     3.5,     6.0,    32.6,    72.3,    74.0,   20.0,  0.18334,      0,      0,       0,       0,       0,       0
total,       1077456,   29968,   29968,   29968,     1.7,     1.6,     2.9,     3.6,     7.0,     8.2,   21.0,  0.17943,      0,      0,       0,       0,       0,       0
total,       1105490,   28034,   28034,   28034,     1.8,     1.8,     3.5,     4.6,     5.3,    13.8,   22.0,  0.17609,      0,      0,       0,       0,       0,       0
total,       1132221,   26731,   26731,   26731,     1.9,     1.8,     3.8,     5.2,     8.4,    11.1,   23.0,  0.17314,      0,      0,       0,       0,       0,       0
total,       1162149,   29928,   29928,   29928,     1.7,     1.7,     3.0,     4.5,     8.0,     9.1,   24.0,  0.16950,      0,      0,       0,       0,       0,       0
...
```

After:
```
type       total ops,    op/s,    pk/s,   row/s,    mean,     med,     .95,     .99,    .999,     max,   time,   stderr, errors,  gc: #,  max ms,  sum ms,  sdv ms,      mb
...
total,        822863,   94379,   94379,   94379,     0.5,     0.3,     2.0,     2.0,     2.1,     3.7,    9.0,  0.06669,      0,      0,       0,       0,       0,       0
total,        937337,  114474,  114474,  114474,     0.4,     0.2,     2.0,     2.0,     2.1,     3.4,   10.0,  0.06301,      0,      0,       0,       0,       0,       0
total,        986630,   49293,   49293,   49293,     1.0,     1.0,     2.0,     2.1,    17.9,    19.0,   11.0,  0.07318,      0,      0,       0,       0,       0,       0
total,       1026734,   40104,   40104,   40104,     1.2,     1.0,     2.0,     2.2,     6.3,     7.1,   12.0,  0.08410,      0,      0,       0,       0,       0,       0
total,       1066124,   39390,   39390,   39390,     1.3,     1.0,     2.0,     2.2,     2.6,     3.4,   13.0,  0.09108,      0,      0,       0,       0,       0,       0
total,       1103082,   36958,   36958,   36958,     1.3,     1.1,     2.1,     2.5,     3.1,     4.2,   14.0,  0.09643,      0,      0,       0,       0,       0,       0
total,       1141987,   38905,   38905,   38905,     1.3,     1.0,     2.0,     2.4,    11.4,    12.7,   15.0,  0.09894,      0,      0,       0,       0,       0,       0
total,       1180023,   38036,   38036,   38036,     1.3,     1.0,     2.0,     3.7,     5.6,     7.1,   16.0,  0.10070,      0,      0,       0,       0,       0,       0
total,       1216481,   36458,   36458,   36458,     1.4,     1.0,     2.1,     3.6,     4.7,     5.0,   17.0,  0.10210,      0,      0,       0,       0,       0,       0
total,       1256819,   40338,   40338,   40338,     1.2,     1.0,     2.0,     2.2,     3.5,     5.4,   18.0,  0.10173,      0,      0,       0,       0,       0,       0
total,       1295122,   38303,   38303,   38303,     1.3,     1.0,     2.0,     2.4,    21.0,    21.1,   19.0,  0.10136,      0,      0,       0,       0,       0,       0
total,       1334743,   39621,   39621,   39621,     1.3,     1.0,     2.0,     2.3,     3.3,     4.0,   20.0,  0.10055,      0,      0,       0,       0,       0,       0
total,       1375579,   40836,   40836,   40836,     1.2,     1.0,     2.0,     2.1,     3.4,     5.7,   21.0,  0.09927,      0,      0,       0,       0,       0,       0
total,       1415576,   39997,   39997,   39997,     1.2,     1.0,     2.0,     2.3,     3.2,     4.1,   22.0,  0.09807,      0,      0,       0,       0,       0,       0
total,       1449268,   33692,   33692,   33692,     1.5,     1.4,     2.5,     3.2,     4.2,     5.6,   23.0,  0.09800,      0,      0,       0,       0,       0,       0
total,       1471873,   22605,   22605,   22605,     2.2,     2.0,     4.8,     5.9,     7.0,     7.9,   24.0,  0.10015,      0,      0,       0,       0,       0,       0
...
```

Fixes: https://github.com/scylladb/scylladb/issues/24411

This is a new feature, so no backport needed.

andrzej-jackowski-scylladb commented on cql3/query_processor.cc:
```diff
@@ -657,6 +657,18 @@ query_processor::do_execute_prepared(
+template<typename Func>
+auto maybe_change_sl_to_driver(const ::shared_ptr<cql_statement>& statement, service::query_state& query_state, Func&& fn)
+{
+    if (statement->depends_on("system_schema", std::optional<std::string_view>{}) || statement->depends_on("system", "peers") || statement->depends_on("system", "local")) {
```
I believe this line requires more attention before moving this PR as "ready for review". Do we really want to move all "system_schema" load to `sl:driver`? Maybe, at least, we should limit to SELECT statements (as those are initiated by drivers)

andrzej-jackowski-scylladb replied:
Modified the mechanism to work as explained in https://github.com/scylladb/scylla-drivers/issues/80#issuecomment-3197349782

ptrsmrn commented on db/system_keyspace.cc:
```diff
@@ -2908,6 +2908,22 @@ future<mutation> system_keyspace::make_view_builder_version_mutation(api::timest
     co_return std::move(muts[0]);
 }
 
+static constexpr auto SERVICE_LEVEL_DRIVER_CREATED_KEY = "service_level_driver_created";
```
Consider shorter form: SL_DRIVER_CREATED_KEY = sl_driver_created

ptrsmrn commented on service/qos/service_level_controller.cc:
```diff
@@ -385,6 +385,31 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
 
+future<> service_level_controller::create_driver_service_level(service::raft_group0_client& group0_client, db::system_keyspace& sys_ks) {
+    try {
+        // Users shouldn't remove sl:driver, but if they do, don't recreate it
+        std::optional<bool> driver_service_level_already_created = co_await sys_ks.get_service_level_driver_created();
+        if (driver_service_level_already_created.has_value() && driver_service_level_already_created.value()) {
+            co_return;
+        }
```
The comments says we won't recreate a removed SL. What about the case where `driver_service_level_already_created` is set to `false` (so has a value and this value is false) - how can we end up here? By some internal failure? If user deletes this SL and sets this value to false, would we recreate it despite his intentions?

andrzej-jackowski-scylladb replied:
It is possible to manually change `driver_service_level_already_created=false` using `UPDATE ... ` query.
After that, `sl:driver` will be recreated if it was removed. I think that if `driver_service_level_already_created` is set to false it means that the intention is to actually re-create it.

I can modify the check to just verify the existence of `driver_service_level_already_created`. I decided to also check the value, but I can remove this logic if it brings more confusion than clearness.

ptrsmrn commented on docs/features/workload-prioritization.rst:
```diff
@@ -23,9 +23,13 @@ By assigning each service level to the different roles within your organization,
 .. _`role` : /operating-scylla/security/rbac_usecase/
 
+Initially ScyllaDB has two service levels:
+ - default that receives all load unassigned to any other service level
+ - driver that is used for fetching metadata required by drivers and for connection creation before user is authenticated and authorized
```
I suggest to somehow emphasize that "default" and "driver" are SL names. Just surrounding them with quotes could be enough.

andrzej-jackowski-scylladb replied:
Fixed

ptrsmrn commented on docs/features/workload-prioritization.rst:
```diff
@@ -23,9 +23,13 @@ By assigning each service level to the different roles within your organization,
 
 .. _`role` : /operating-scylla/security/rbac_usecase/
 
+Initially ScyllaDB has two service levels:
```
@Annamikhlin - can you please go over the docs change and possbily approve?

Annamikhlin replied:
@annastuchlik - now the correct Anna :)

ptrsmrn replied:
sorry and thanks :)

mykaul replied:
@annastuchlik ?

andrzej-jackowski-scylladb commented on docs/features/workload-prioritization.rst:
```diff
@@ -76,7 +80,7 @@ Where:
 .......
 
-There are 3 service levels (OLAP, OLTP, Default) where: (the percentage of resources = (Assigned Shares / Total Shares) x 100). Total Shares in this case is the total of all allocated shares + the Default SLA (1000). The percentage of resources would be:
+There are 3 service levels (OLAP, OLTP, driver, default) where: (the percentage of resources = (Assigned Shares / Total Shares) x 100). Total Shares in this case is the total of all allocated shares + the Default SLA (1000). The percentage of resources would be:
```
Fixed.

avikivity commented on docs/features/workload-prioritization.rst:
```diff
@@ -23,9 +23,13 @@ By assigning each service level to the different roles within your organization,
+Initially ScyllaDB has two service levels:
+ - ``'default'`` that receives all load unassigned to any other service level
+ - ``'driver'`` that is used for fetching metadata required by drivers and for connection creation before user is authenticated and authorized
+
```
Please add some words about how the driver service level helps the user (preventing metadata fetches during reconnects from overloading) and about how the user should manage the driver service level (don't touch it).

andrzej-jackowski-scylladb replied:
Added.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -2908,6 +2908,22 @@ future<mutation> system_keyspace::make_view_builder_version_mutation(api::timest
+
+future<mutation> system_keyspace::make_service_levels_driver_created_mutation(bool is_created, api::timestamp_type timestamp) {
+    static sstring query = format("INSERT INTO {}.{} (key, value) VALUES (?, ?);", db::system_keyspace::NAME, db::system_keyspace::SCYLLA_LOCAL);
+    auto muts = co_await _qp.get_mutations_internal(query, internal_system_query_state(), timestamp, {SERVICE_LEVEL_DRIVER_CREATED_KEY, format("{}", is_created)});
```
Ah, I see get_scylla_local_param_as() isn't completely custom, it uses abstract_type serialization. And so should you here, convert is_created to a data_value and convert that to text.

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -385,6 +385,36 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
+    while (!_global_controller_db->group0_aborter.abort_requested()) {
+        try {
+            // Users shouldn't remove sl:driver, but if they do, don't recreate it
+            std::optional<bool> driver_service_level_already_created = co_await sys_ks.get_service_level_driver_created();
```
Shouldn't this be in the transaction? copying a group0 guru. @patjed41

piodul replied:
The thing is, the `create_driver_service_level` function is being called as a part of the startup sequence. If we require `start_operation` to successfully execute during node startup then I suppose we might encounter problems if we try to restart nodes when we lose raft quorum. Additionally, I have doubts that this will work in group0 recovery mode.

Perhaps `create_driver_service_level` could be called by the topology coordinator?

piodul replied:
Ah, ok, I see that we only attempt doing it once.

In order to avoid problems, perhaps we could do double-checked locking: check it before taking the guard, start_operation (which acquires the guard), then check again.

avikivity replied:
I'd be happier if it's done from the topology change coordinator.

andrzej-jackowski-scylladb replied:
Moved `sl:driver` creating logic to `topology_coordinator`

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -385,6 +385,36 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
+            slo.workload = service_level_options::workload_type::batch;
+
+            auto guard = co_await group0_client.start_operation(_global_controller_db->group0_aborter);
+            auto status_mut = co_await sys_ks.make_service_levels_driver_created_mutation(true, guard.write_timestamp());
```
Could race here.

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -385,6 +385,36 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
+            auto status_mut = co_await sys_ks.make_service_levels_driver_created_mutation(true, guard.write_timestamp());
+
+            service::group0_batch mc{std::move(guard)};
+            bool if_not_exsists = true;
```
spelling

andrzej-jackowski-scylladb replied:
fixed

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -385,6 +385,36 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
+            mc.add_mutation(status_mut, "set service_level_driver_created=true");
+            co_return co_await commit_mutations(std::move(mc));
+        } catch (service::group0_concurrent_modification&) {
+            continue; // Let's retry
```
If something goes very badly, this will spin, so abortable_sleep here.

andrzej-jackowski-scylladb replied:
Removed the loop, as creation of `sl:driver` was moved to `topology_coordinator`

avikivity commented on transport/server.cc:
```diff
@@ -639,6 +639,7 @@ cql_server::connection::connection(cql_server& server, socket_address server_add
     , _client_state(service::client_state::external_tag{}, server._auth_service, &server._sl_controller, server.timeout_config(), addr)
     , _current_scheduling_group(default_scheduling_group())
 {
+    update_scheduling_group_to_driver();
```
Is this needed? Just run the accept loop under the driver service level.

I guess that doesn't work for a listener that was started before sl:driver existed.

andrzej-jackowski-scylladb replied:
Modified the code to switch scheduling group in `server::do_accepts`

avikivity commented:
I'd like the cover letter to contain a description of a manual test that was used to verify not just the inner workings, but that the purpose of the patch was achieved.

1. start a workload (could be a single node single shard cluster)
2. unleash a connection storm, show bad behavior
3. upgrade to new version
4. redo, show good behavior

You can skew the setup to make the bad behavior pronounced, for example create 1000 keyspaces with a table each. I know some drivers will generate O(nr_keyspaces) queries. Have a thousand clients for the connection storm.

piodul commented on db/system_keyspace.cc:
```diff
@@ -2908,6 +2908,22 @@ future<mutation> system_keyspace::make_view_builder_version_mutation(api::timest
 
+static constexpr auto SERVICE_LEVEL_DRIVER_CREATED_KEY = "service_level_driver_created";
+
+future<std::optional<bool>> system_keyspace::get_service_level_driver_created() {
```
Why `std::optional<bool>`? Can't we flatten it and treat `null` as `false` here?

avikivity replied:
It's better for a low-level accesor not to drop information IMO. The caller can collapse it with optional::value_or() if it wants.

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -385,6 +385,36 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
+            service::group0_batch mc{std::move(guard)};
+            bool if_not_exsists = true;
+            co_await add_distributed_service_level(service_level_controller::driver_service_level_name, slo, if_not_exsists, mc);
+            mc.add_mutation(status_mut, "set service_level_driver_created=true");
```
It would also be good to create the "driver" service level immediately when creating a new cluster. Look for the `build_mutation_from_join_params` function.

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -483,6 +483,17 @@ future<foreign_ptr<std::unique_ptr<cql_server::response>>>
+                client_state.set_sl_state(service::client_state::service_level_state::CONTROL_CONNECTION);
+            } else if (cqlop == cql_binary_opcode::QUERY || cqlop == cql_binary_opcode::PREPARE || cqlop == cql_binary_opcode::EXECUTE || cqlop == cql_binary_opcode::BATCH) {
+                client_state.set_sl_state(service::client_state::service_level_state::USER);
+                update_scheduling_group_to_user();
```
Added `co_await coroutine::switch_to` just after `update_scheduling_group_to_user` to address the case

andrzej-jackowski-scylladb commented on db/system_keyspace.cc:
```diff
@@ -2908,6 +2908,22 @@ future<mutation> system_keyspace::make_view_builder_version_mutation(api::timest
 
+static constexpr auto SERVICE_LEVEL_DRIVER_CREATED_KEY = "service_level_driver_created";
+
+future<std::optional<bool>> system_keyspace::get_service_level_driver_created() {
```
Added the explanation to the commit message

andrzej-jackowski-scylladb commented on main.cc:
```diff
@@ -2292,6 +2292,7 @@ sharded<locator::shared_token_metadata> token_metadata;
             // update the service level cache after the SL data accessor and auth service are initialized.
             if (sl_controller.local().is_v2()) {
                 sl_controller.local().update_cache(qos::update_both_cache_levels::yes).get();
+                sl_controller.local().create_driver_service_level(group0_client, sys_ks.local()).get();
```
Added `DRIVER_SERVICE_LEVEL` feature

andrzej-jackowski-scylladb commented:
v6:
 - Updated `process_request_one` to check the current scheduling group, to avoid race condition between setting `service::client_state::service_level_state::USER` and running requests in parallel, just after the service level was changed.

andrzej-jackowski-scylladb commented:
v7:
 - Rebased to remove conflicts

andrzej-jackowski-scylladb commented on service/qos/service_level_controller.cc:
```diff
@@ -385,6 +385,29 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
+
+        auto status_mut = co_await sys_ks.make_service_levels_driver_created_mutation(true, guard.write_timestamp());
+        service::group0_batch mc{std::move(guard)};
+        bool if_not_exists = true;
```
fixed

andrzej-jackowski-scylladb commented:
v8:
 - Cosmetic changes to resolve review comments

andrzej-jackowski-scylladb commented:
v9:
 - Added `sl:driver` creation in `raft_initialize_discovery_leader` to increase the feature reliability (as proposed by @nuivall to minimize cases when for some reason `sl:driver is not created) and also make eliminate tests flakiness.

I also analyzed the recent test failures:

- [test_one_big_mutation_corrupted_on_startup.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19428/testReport/junit/cluster.dtest.commitlog_test/TestCommitLog/dev___test_py___dev___test_one_big_mutation_corrupted_on_startup_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_one_big_mutation_corrupted_on_startup.dev.3) is https://github.com/scylladb/scylladb/issues/25627
- [test_mv_retried_writes_reach_all_replicas.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19428/testReport/junit/cluster.mv/test_mv_admission_control/debug___test_py___debug___test_mv_retried_writes_reach_all_replicas_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.1) and [test_view_building_with_tablet_move.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19428/testReport/junit/cluster.mv/test_mv_building/debug___test_py___debug___test_view_building_with_tablet_move_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_building_with_tablet_move.debug.3) is https://github.com/scylladb/scylladb/issues/25702
 - `sla_per_user_test.*` failures are expected (just need to adjust the check for the existence of `sl:driver`, I will update dtests when the PR will be ready.
 - All other test failures I expect will be fixed by the `sl:driver` creation in `raft_initialize_discovery_leader`

andrzej-jackowski-scylladb commented:
v10:
 - `test_describe.py` - prevent from removing `sl:driver` in test cleanup, it broke other cqlpy tests
- Comment update in `test_service_levels.py`

> #### 🔴 CI State: FAILURE
I reviewed the recent test failures:
 - `cqlpy.test_service_levels.test_scheduling_groups_limit` failed because the node state was invalid due to `test_describe.py` cleanup (fixed in v10)
 - `sla_per_user_test.TestSLA*`  were expected to fail (dtests need adjustment for the new service level)
 - `enterprise_upgrade_test.TestEnterpriseUpgradeFromOSS.test_workload_prioritization_after_upgrade` dtest needs similar fix as `sla_per_user_test`

andrzej-jackowski-scylladb commented on generic_server.cc:
```diff
@@ -384,6 +386,7 @@ future<> server::do_accepts(int which, bool keepalive, socket_address server_add
                 }
             }
             accept_result cs_sa = co_await _listeners[which].accept();
+            co_await coroutine::switch_to(get_scheduling_group());
```
Ok, removed it

andrzej-jackowski-scylladb commented on test/cluster/auth_cluster/test_raft_service_levels.py:
```diff
@@ -553,3 +553,20 @@ async def test_driver_service_creation_failure(manager: ManagerClient) -> None:
+    await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
+    metrics = await manager.metrics.query(server.ip_addr)
+
+    assert get_processed_tasks_for_group(metrics, 'sl:driver') > 0
```
I checked that the server startup sequence already uses `sl:driver`. I just removed the entire tests, as there are other, more detailed tests checking `get_processed_tasks_for_group(metrics, 'sl:driver')`

andrzej-jackowski-scylladb commented on test/cluster/auth_cluster/test_raft_service_levels.py:
```diff
@@ -496,3 +496,60 @@ async def test_reload_service_levels_after_auth_service_is_stopped(manager: Mana
+        await cql.run_async(f"CREATE SERVICE_LEVEL sl_{i}", host=h)
+    service_levels = await cql.run_async("LIST ALL SERVICE LEVELS")
+    assert len(service_levels) == (max_user_service_levels + 1)
+    for sl in service_levels:
```
fixed

andrzej-jackowski-scylladb commented on test/cluster/auth_cluster/test_raft_service_levels.py:
```diff
@@ -496,3 +496,60 @@ async def test_reload_service_levels_after_auth_service_is_stopped(manager: Mana
+    await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
+
+    await cql.run_async(f"DROP SERVICE LEVEL driver", host=h)
+    max_user_service_levels = 7
```
fixed

andrzej-jackowski-scylladb commented on test/cluster/auth_cluster/test_raft_service_levels.py:
```diff
@@ -496,3 +496,60 @@ async def test_reload_service_levels_after_auth_service_is_stopped(manager: Mana
+    assert len(await cql.run_async("LIST ALL SERVICE LEVELS")) == 0
+
+@pytest.mark.asyncio
+async def test_driver_service_creation_failure(manager: ManagerClient) -> None:
```
I reconsidered it, however, the purpose of the test is testing the creation failure, and the way it is achieved is kind of a detail (i.e. we don't have creation_failure tests that achieves the same differently). Therefore, I think the current naming is more relevant

andrzej-jackowski-scylladb commented on generic_server.cc:
```diff
@@ -15,6 +15,7 @@
 #include <seastar/core/reactor.hh>
```
updated.

andrzej-jackowski-scylladb commented:
v14:
 - Added special handling of `sl:driver` for `DESC ...`, and `test_desc_driver_service_level`
 - Improved `transport/server.cc` logic to get rid of `co_return co_await ` and increase code clearness

v15:
 - Empty commit, just fixed commit author problem introduced with git rebase

andrzej-jackowski-scylladb commented:
v16:
 - Reworked the service level changing logic in `transport/server.cc` because the v14 implementation added ~200 instructions on critical path (it was just a single if, but accessing `static thread_local` variable via `current_scheduling_group()` and comparing sg names).
 - Fixes for `cqlpy.test_service_levels.test_scheduling_groups_limit` and `cluster.auth_cluster.test_raft_service_levels.test_service_levels_upgrade`


**Micro benchmarks of transport/server**
As stated above, v14 implementation introduced an overhead of ~200 instructions on handling `QUERY` requests, measured by the [perf_cql_raw](https://github.com/scylladb/scylladb/commit/67fbe35833e2d23a8e9c2dcb5e04580231d8ec96) benchmark.

Baseline (6116f9e11b4c00de60c2e8b11b93edb66c39a3b2):
```
instructions_per_op:
	mean=   108075.64 standard-deviation=633.52
	median= 108011.68 median-absolute-deviation=64.79
	maximum=114346.67 minimum=107996.44
```

v14 (11849360966460067a4282388f24db724b5b7f1b):
```
instructions_per_op:
	mean=   108235.52 standard-deviation=662.70
	median= 108166.23 median-absolute-deviation=70.11
	maximum=114795.26 minimum=108144.00
```

Final version (e3b9ec070f5cd3536d411665d53c9de689c6ae0a):
```
instructions_per_op:
	mean=   108039.50 standard-deviation=633.56
	median= 107970.67 median-absolute-deviation=70.02
	maximum=114309.34 minimum=107950.68
```

The experiments above were repeated many times, and the results were always similar. The difference was also visible after refreshing pgo profiles.

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -12,6 +12,7 @@
 #include "cql3/statements/modification_statement.hh"
```
There are some metrics that included information about the used scheduling group, like the ones that I used to verify the feature correctness ( see https://github.com/scylladb/scylladb/pull/25412/commits/946e96bf25d24f084b06baaf1bf056c5e42a6d8b)

andrzej-jackowski-scylladb replied:
Changed the logic, now we only have `update_scheduling_group` that updates the tenant unconditionally

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -982,11 +999,14 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
 
 void cql_server::connection::update_to_user_scheduling_group() {
+    if (_client_state.get_service_level_state() == service::client_state::service_level_state::USER 
+            || _client_state.get_service_level_state() == service::client_state::service_level_state::NO_DRIVER_SL) {
```
update_scheduling_group

andrzej-jackowski-scylladb commented:
v17:
 - Only cosmetic code / comments changes

andrzej-jackowski-scylladb commented:
@piodul, @Jadw1, please review the recent version. The following major changes were introduced since your last review:
- Creation of `sl:driver` was moved to `topology_coordinator` and additionally it is created in `raft_initialize_discovery_leader` for new systems and tests.
- Special handling of `DESC ...` was added to make sure `sl:driver` works correctly in backups.
- `transport/server.cc` logic was reworked to reduce complexity and avoid overheads
- Additional tests added to cover the new changes.
- Some (flaky) tests were fixed.

nuivall commented on generic_server.cc:
```diff
@@ -366,6 +367,7 @@ server::listen(socket_address addr, std::shared_ptr<seastar::tls::credentials_bu
 
 future<> server::do_accepts(int which, bool keepalive, socket_address server_addr) {
     while (!_gate.is_closed()) {
+        co_await coroutine::switch_to(get_scheduling_group());
```
I am not sure what this is supposed to do, it calls seastar's current_sheduling_group, when changing to current has any effect?

nuivall commented on service/qos/service_level_controller.cc:
```diff
@@ -407,6 +407,47 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
     }
 }
 
+service_level_options get_driver_service_level_slo() {
```
static?

andrzej-jackowski-scylladb replied:
Definitely, will fix it

nuivall commented on service/qos/raft_service_level_distributed_data_accessor.hh:
```diff
@@ -41,6 +41,7 @@ public:
     virtual bool is_v2() const override;
     virtual bool can_use_effective_service_level_cache() const override;
     virtual ::shared_ptr<service_level_distributed_data_accessor> upgrade_to_v2(cql3::query_processor& qp, service::raft_group0_client& group0_client) const override;
+    static future<std::pair<utils::chunked_vector<mutation>, sstring>> set_service_level_mutations(cql3::query_processor& qp, sstring service_level_name, qos::service_level_options slo, api::timestamp_type timestamp);
```
I am not sure if it is worth to complicate function return type so much only for debugging purposes, while sstring is actually hardcoded

nuivall replied:
maybe just skip the whole cql statement in description, just logically write what those mutations are doing

andrzej-jackowski-scylladb commented on service/qos/raft_service_level_distributed_data_accessor.cc:
```diff
@@ -56,30 +56,33 @@ static void validate_state(const service::raft_group0_client& group0_client) {
             : data_value(qos::service_level_options::to_string(slo.workload));
 
-    auto muts = co_await _qp.get_mutations_internal(insert_query, qos_query_state(), mc.write_timestamp(), {service_level_name, timeout_to_data_value(slo.timeout), workload});
+    auto muts = co_await qp.get_mutations_internal(insert_query, qos_query_state(), timestamp, {service_level_name, timeout_to_data_value(slo.timeout), workload});
```
`set_service_level_mutations` need to be static, so we don't have the access to `_qp`

nuivall commented on service/qos/service_level_controller.cc:
```diff
@@ -407,6 +407,47 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
+    co_return mutations;
+}
+
+future<> service_level_controller::create_driver_service_level(service::group0_guard guard, db::system_keyspace& sys_ks) {
```
I'd call it: migrate_to_xxxx to show this is not the normal path, also shouldn't it reuse service_level_controller::get_create_driver_service_level_mutations

nuivall commented on service/qos/service_level_controller.hh:
```diff
@@ -104,6 +105,7 @@ using update_both_cache_levels = bool_class<class update_both_cache_levels_tag>;
 class service_level_controller : public peering_sharded_service<service_level_controller>, public service::endpoint_lifecycle_subscriber {
 public:
     static inline const int32_t default_shares = 1000;
+    static constexpr std::string driver_service_level_name = "driver";
```
for consistency maybe std::string -> sstring

nuivall commented on service/qos/service_level_controller.hh:
```diff
@@ -344,6 +346,15 @@ public:
     void maybe_start_legacy_update_from_distributed_data(std::function<steady_clock_type::duration()> interval_f, service::storage_service& storage_service, service::raft_group0_client& group0_client);
 
+    /**
+     * Get mutations required to create `sl:driver` and remember it in `system.scylla_local`
```
Why this ` and remember it in `system.scylla_local`` part? This function is not doing that.

nuivall commented on service/qos/service_level_controller.hh:
```diff
@@ -344,6 +346,15 @@ public:
+     */
+    future<utils::chunked_vector<mutation>> get_create_driver_service_level_mutations(db::system_keyspace& sys_ks, api::timestamp_type timestamp) const;
+    /**
+     * Create `sl:driver` using _sl_data_accessor if possible, and remember it in `system.scylla_local`
```
I'd use `store` instead of remember

nuivall commented on service/storage_service.cc:
```diff
@@ -1357,6 +1358,11 @@ future<> storage_service::raft_initialize_discovery_leader(const join_node_reque
 
     insert_join_request_mutations.emplace_back(co_await _sys_ks.local().make_auth_version_mutation(write_timestamp, db::system_keyspace::auth_version_t::v2));
 
+    auto sl_driver_mutations = co_await _sl_controller.local().get_create_driver_service_level_mutations(_sys_ks.local(), write_timestamp);
```
perhaps get_create_driver_service_level_mutations should be static too, if set_service_level_mutations is static now

dawmd commented on service/qos/service_level_controller.cc:
```diff
@@ -1033,6 +1040,34 @@ future<std::vector<cql3::description>> service_level_controller::describe_create
+            .create_statement = managed_string(alter_statement)
+        });
+    } else {
+        const sstring drop_statement = describe_service_level(driver_service_level_name, service_level_options{}, "DROP SERVICE LEVEL IF EXISTS");
```
I'm a bit confused. I'm not reviewing the whole PR, so this might be an irrelevant comment. Feel free to ignore it in that case. :)

If I remember the design document correctly, one of the requirements was to disallow dropping the driver service level. Since we have this piece of code, I guess my memory is iffy, so I'll assume that it's possible to drop it.

In that case, we should make sure that the driver service level always comes first in the output of DESCRIBE. Why? If the user drops it, then they have one additional slot for a custom service level. Assume that that's what happens, i.e.

1. The user drops the driver service level.
2. The user creates service levels, filling in every spot, including the one previously belonging to the driver SL.
3. The user performs a backup.
4. We get a fresh cluster and restore the schema.
5. Other service level statements are executed, but the last fails because there's no more room for it (the driver service level is created automatically).
6. Inconsistent schema.

dawmd replied:
It would be nice to have a test for it too if it's really a valid scenario.

andrzej-jackowski-scylladb replied:
It is a very valid scenario. Fixed it and implemented a test that verifies it.

andrzej-jackowski-scylladb commented on service/qos/service_level_controller.hh:
```diff
@@ -345,6 +346,18 @@ public:
+     * 1. create `sl:driver`
+     * 2. store information that `sl:driver` was created in `system.scylla_local`
+     */
+    static future<utils::chunked_vector<mutation>> get_create_driver_service_level_mutations(db::system_keyspace& sys_ks, api::timestamp_type timestamp);
```
We are calling it from `storage service` on an existing `sl_controller`, however,  `_sl_data_accessor` is still unavailable (`reload_distributed_data_accessor` is called in `main.cc` after `join_cluster`).

Therefore, I made the method static to avoid a misuse (e.g. calling the non-existing `_sl_data_accessor`). If it is a problem, I can make it non-static.

andrzej-jackowski-scylladb commented on test/cqlpy/test_describe.py:
```diff
@@ -2828,6 +2828,52 @@ def test_desc_auth_attach_service_levels(cql, scylla_only):
+    the system. Therefore, it requires special handling in DESC SCHEMA WITH INTERNALS,
+    to make sure that:
+     1. CREATE SERVICE LEVEL doesn't fail if the service level already exists (i.e. IF NOT EXISTS to CREATE)
+     2. If service level exists, it's configuration is fully restored (i.e. ALTER SERVICE command is also added)
```
fixed

andrzej-jackowski-scylladb commented on service/qos/service_level_controller.cc:
```diff
@@ -1033,6 +1040,34 @@ future<std::vector<cql3::description>> service_level_controller::describe_create
+            .create_statement = managed_string(drop_statement)
+        });
+    }
+
```
Implemented.

andrzej-jackowski-scylladb commented on test/cqlpy/test_describe.py:
```diff
@@ -2828,6 +2828,52 @@ def test_desc_auth_attach_service_levels(cql, scylla_only):
+        assert drop.create_statement == "DROP SERVICE LEVEL IF EXISTS driver;"
+
+        # Bring the system to back the initial state for other tests
+        cql.execute("CREATE SERVICE LEVEL DRIVER WITH WORKLOAD_TYPE = 'batch' AND SHARES = 200")
```
Moved to `AuthSLContext`

andrzej-jackowski-scylladb commented on test/cqlpy/test_describe.py:
```diff
@@ -2828,6 +2828,52 @@ def test_desc_auth_attach_service_levels(cql, scylla_only):
+        assert drop.create_statement == "DROP SERVICE LEVEL IF EXISTS driver;"
+
+        # Bring the system to back the initial state for other tests
+        cql.execute("CREATE SERVICE LEVEL DRIVER WITH WORKLOAD_TYPE = 'batch' AND SHARES = 200")
```
fixed

andrzej-jackowski-scylladb commented on generic_server.cc:
```diff
@@ -366,6 +367,7 @@ server::listen(socket_address addr, std::shared_ptr<seastar::tls::credentials_bu
 
 future<> server::do_accepts(int which, bool keepalive, socket_address server_addr) {
     while (!_gate.is_closed()) {
+        co_await coroutine::switch_to(get_scheduling_group_for_new_connection());
```
My team asked me to remove on of two uses of `switch_to` for code clearity, but I agree it's more correct if there are two calls, just as you proposed. Implemented your idea.

andrzej-jackowski-scylladb commented on test/cluster/auth_cluster/test_raft_service_levels.py:
```diff
@@ -498,3 +498,70 @@ async def test_reload_service_levels_after_auth_service_is_stopped(manager: Mana
+    assert "driver" not in service_level_names
+
+    logger.info("Set service_level_driver_created=false in system.scylla_local and restart to force re-creation of sl:driver")
+    await cql.run_async(f"UPDATE system.scylla_local SET value = 'false' WHERE key  = 'service_level_driver_created'", host=h)
```
ok, change the code to call it on all nodes

andrzej-jackowski-scylladb commented on service/qos/service_level_controller.cc:
```diff
@@ -407,6 +407,47 @@ future<> service_level_controller::update_cache(update_both_cache_levels update_
+    std::move(sl_mutations.begin(), sl_mutations.end(), std::back_inserter(mutations));
+
+    auto sys_ks_mutation = co_await sys_ks.make_service_level_driver_created_mutation(true, timestamp);
+    mutations.push_back(sys_ks_mutation);
```
Added move.

dawmd commented on test/cqlpy/test_describe.py:
```diff
@@ -14,7 +14,7 @@
 from .util import new_type, unique_name, new_test_table, new_test_keyspace, new_function, new_aggregate, \
     new_cql, keyspace_has_tablets, unique_name_prefix, new_session, new_user, new_materialized_view, \
-    new_secondary_index
+    new_secondary_index, get_max_user_service_levels
```
[commit: `test: add test_desc_driver_service_level`]
super nitpicky: The title and description of the commit are out-of-date (they only mention one of the tests).

dawmd commented on transport/server.cc:
```diff
@@ -973,6 +973,7 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
             res = make_autheticate(stream, a.qualified_java_name(), trace_state);
         }
     } else {
+        update_scheduling_group();
```
Should this be moved to a separate PR? Or is it necessary for adding the driver service level?

andrzej-jackowski-scylladb replied:
Without this we would leave the unauthorized connection in `sl:driver` state. I think, if possible, it's better to include it here (or at least merge it before this PR).

dawmd commented on transport/server.cc:
```diff
@@ -982,11 +982,11 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
-        _current_scheduling_group = shg;
-        co_return co_await _server._sl_controller.with_user_service_level(_client_state.user(), std::move(process_loop));
-    });
+        switch_tenant([this] (noncopyable_function<future<> ()> process_loop) -> future<> {
```
[commit: ` transport: whitespace only change in update_scheduling_group`]
nit: For better review experience, I think it would be better to have this commit right before the changes that need it.

andrzej-jackowski-scylladb replied:
fixed.

andrzej-jackowski-scylladb commented:
> [test_view_building_with_tablet_move.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20011/testReport/junit/cluster.mv/test_mv_building/debug___test_py___debug___test_view_building_with_tablet_move_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_building_with_tablet_move.debug.1)

https://github.com/scylladb/scylladb/issues/25804

andrzej-jackowski-scylladb commented on test/cqlpy/test_describe.py:
```diff
@@ -2831,6 +2831,70 @@ def test_desc_auth_attach_service_levels(cql, scylla_only):
+    Driver service level is a special service level that is created automatically by
+    the system. Therefore, it requires special handling in DESC SCHEMA WITH INTERNALS,
+    to make sure that:
+     1. CREATE SERVICE LEVEL doesn't fail if the service level already exists (i.e. IF NOT EXISTS to CREATE)
```
I improved the comments and I hope that now the changes in `test_describe.py` are clear.

`sl:driver` requires a very special handling in `DESC SCHEMA WITH INTERNALS`. We emit either a pair of `CREATE SERVICE LEVEL IF NOT EXISTS` + `ALTER SERVICE LEVEL` or `DROP SERVICE LEVEL IF EXISTS`. All other service levels emits just a single `CREATE SERVICE LEVEL` per service level.

andrzej-jackowski-scylladb commented on test/cqlpy/test_service_level_api.py:
```diff
@@ -134,6 +141,8 @@ def test_switch_tenants(cql):
             wait_for_clients(cql, user, 3) # 3 from smp=2 + control connection
             wait_until_all_connections_authenticated(cql)
+            send_some_requests(cql)
+            send_some_requests(user_session)
```
Those changes in test/cqlpy/test_service_level_api.py were needed to trigger the change from sl:driver to user's service level. They are no longer needed, because the implementation was changed in v16 (ref. https://github.com/scylladb/scylladb/pull/25412#issuecomment-3274258922). I think we are not going back to the previous implementation, so I reverted them.

andrzej-jackowski-scylladb commented on test/cqlpy/test_describe.py:
```diff
@@ -2831,6 +2831,70 @@ def test_desc_auth_attach_service_levels(cql, scylla_only):
+
+    with AuthSLContext(cql):
+        cql.execute(f"ALTER SERVICE LEVEL driver WITH SHARES=123")
+        cql.execute(f"ALTER SERVICE LEVEL driver WITH TIMEOUT=321s")
```
I changed the implementation of `AuthSLContext` - now it obtains the `sl:driver` configuration in `__enter__` and restores it in `__exit__`. Therefore, altering `sl:driver` in `AuthSLContext` will not affect other tests anymore.

andrzej-jackowski-scylladb commented on test/cqlpy/util.py:
```diff
@@ -373,3 +373,6 @@ def match_kv(kv):
         return result
+
+def get_max_user_service_levels():
+    return 7
```
This discussion was continued in https://github.com/scylladb/scylladb/pull/25412#discussion_r2353333622.

andrzej-jackowski-scylladb commented on test/cqlpy/test_describe.py:
```diff
@@ -2831,6 +2831,70 @@ def test_desc_auth_attach_service_levels(cql, scylla_only):
+
+# Marked as `scylla_only` because we verify that the output of `DESCRIBE SCHEMA`
+# contains information about service levels. That's not the case in Cassandra.
+def test_desc_removed_driver_service_level(cql, scylla_only):
```
Added comments with description

andrzej-jackowski-scylladb commented on service/qos/service_level_controller.cc:
```diff
@@ -958,7 +958,27 @@ future<> service_level_controller::unregister_subscriber(qos_configuration_chang
+    DROP_IF_EXISTS,
+};
+
+sstring describe_cmd_to_sstring(describe_cmd cmd_enum) {
```
fixed

andrzej-jackowski-scylladb commented:
v19:
 - Only commit messages / comments fixes

andrzej-jackowski-scylladb commented:
> #### 🟢 CI State: SUCCESS

@piodul CI is green, and I addressed all reviewers remarks. Please review

nyh commented on test/cqlpy/test_service_levels.py:
```diff
@@ -161,7 +161,7 @@ def test_scheduling_groups_limit(scylla_only, cql):
 
     assert created_count > 0
-    assert created_count == 7 # regression check
+    assert created_count == get_max_user_service_levels() # regression check
```
> The tests need ScyllaDB in a state where no more service levels can be created (to test corner cases). 

I still don't understand (maybe I'm just missing something) why we ever needed more than one test to test this specific corner case, of trying to define 7 service levels (needs to be supported because the documentation guarantees it is supported) or how 8 service levels fails. But beyond one test testing this scenario - why do any other service-level test actually need to use specifically 7 of them?

andrzej-jackowski-scylladb replied:
Here is a short description for each test that uses it:
- `test_service_levels_over_limit` in `test/cluster/auth_cluster/test_raft_service_levels.py` - (it's an old test) checks a corner case when after migrating from old OSS we have more SLs that actually permitted

- `test_driver_service_creation_failure` in `test/cluster/auth_cluster/test_raft_service_levels.py` - (it's a new test) uses the limit of service level to trigger failure in creation of `sl:driver` to test the error handling

- `test_desc_removed_driver_service_level` in `test/cqlpy/test_describe.py` - (it's a new test) checks that if driver service level is dropped and all SL slots are occupied, `DESC ...` returns schema that can be applied (this was actually a bug in this PR, see https://github.com/scylladb/scylladb/pull/25412#discussion_r2349048470)
